lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
 using namespace swift;
 
 //===----------------------------------------------------------------------===//
-// NameBinder
+// MARK: NameBinder and supporting types
 //===----------------------------------------------------------------------===//
 
 using ImportedModule = ModuleDecl::ImportedModule;
+using ImportedModuleDesc = SourceFile::ImportedModuleDesc;
 using ImportOptions = SourceFile::ImportOptions;
+using ImportFlags = SourceFile::ImportFlags;
+
+namespace {
+  /// Represents an import which the NameBinder knows exists, but which has not
+  /// yet had its options checked, module loaded, or cross-imports found.
+  ///
+  /// An UnboundImport may represent a physical ImportDecl written in the
+  /// source, or it may represent a cross-import overlay that has been found and
+  /// needs to be loaded.
+  struct UnboundImport {
+    /// If this UnboundImport directly represents an ImportDecl, the ImportDecl
+    /// it represents.
+    ///
+    /// This property should only be used to control where information is added
+    /// to the AST. Don't retrieve information directly from the ImportDecl; use
+    /// the other member variables.
+    NullablePtr<ImportDecl> ID;
+
+    SourceLoc importLoc;
+    ImportOptions options;
+    StringRef privateImportFileName;
+    Located<ImportKind> importKind;
+    ModuleDecl::AccessPathTy modulePath;
+    ModuleDecl::AccessPathTy declPath;
+
+    NullablePtr<DeclAttributes> attrs;
+
+    /// Create an UnboundImport for a user-written import declaration.
+    explicit UnboundImport(ImportDecl *ID);
+
+    /// Make sure the import is not a self-import.
+    bool checkNotTautological(const SourceFile &SF);
```

Comment:
It may not be obvious what "tautological" means in this context (e.g. one might think that you can't be allowed to import the same module because then you could write something like `@_implementationOnly import MyName` which is impossible.). Maybe `checkNotSelfImport`?

Reply:
Some self-imports are allowed, so I prefer this terminology. I'll expand the doc comment to explain:

```
    /// Diagnoses if the import would simply load the module \c SF already
    /// belongs to, with no actual effect.
    ///
    /// Some apparent self-imports do actually load a different module; this
    /// method allows them.
```